### ISSUE TITLE
Spike timings monitor does not work with compile_mode

### ISSUE THREAD

**Issue by davestanley** [DynaSim/DynaSim#546]
Got the following error when trying to compile my code for tracking spike times

```
??? Function 'exist' is not supported for code generation. Consider adding coder.extrinsic('exist') at the top of the function to bypass code
generation.

Error in ==> solve_ode_20181216231302_834 Line: 165 Column: 7
Code generation failed: View Error Report
Error: 
	 in codegen (line 0)
	 in makeMex (line 126)
	 in dsPrepareMEX (line 59)
	 in dsGetSolveFile (line 206)
	 in dsSimulate/tryFn (line 981)
	 in dsSimulate (line 827)
```

The offending section of my solve_ode file is:

```
 % ------------------------------------------------------------
  % Conditional actions:
  % ------------------------------------------------------------
  conditional_test=(I_V(n,:)>=p.I_thresh&I_V(n-1,:)<p.I_thresh);
  if ~exist('I_V_spikes','var')
    I_V_spikes = [];
  end;
  if any(conditional_test), I_V_spikes(n,conditional_test)=1;inds=find(conditional_test); for j=1:length(inds), i=inds(j); I_tspike(I_buffer_index(i),i)=t; I_buffer_index(i)=mod(-1+(I_buffer_index(i)+1),2)+1; end; end

```

Need to remove the if ~exist statement, or find some work around

**Comment by davestanley** [comment]
The following code will reproduce this error. This is just running the default dsDemos code, with mex_flag turned on

```
%% Networks of integrate-and-fire neurons with axonal delays and refractory period
% tau [ms]: membrane time constant (RC)
% tref [ms]: absolute refractory period
% delay [ms]: axonal delay
% tspike: reserved variable for storing past spike times when monitoring spikes

% 1 E-cell driving 1 I-cell
LIF={
    'dV/dt=(E-V+R*I+noise*randn-@isyn)/tau; V(0)=-65'
    'if(any(t<tspike+tref,1))(V=reset)'
    'tau=10; tref=10; E=-70; thresh=-55; reset=-75; R=9; I=1.55; noise=100'
    'monitor V.spikes(thresh)'
     };

iampa={
  'gAMPA=.5; EAMPA=0; tauD=2; tauR=0.4; delay=15'
  'f(x) = (exp(-x/tauD)-exp(-x/tauR)).*(x>0)'
  'Isyn(V) = gAMPA.*sum(f(t-tspike_pre-delay)).*(V-EAMPA)'
  '@isyn += Isyn(V_post)'
};

s=[];
s.populations(1).name='E';
s.populations(1).equations=LIF;
s.populations(2).name='I';
s.populations(2).equations=LIF;
s.populations(2).parameters={'I',0,'noise',0};
s.connections(1).direction='E->I';
s.connections(1).mechanism_list='iampa';
s.mechanisms(1).name='iampa';
s.mechanisms(1).equations=iampa;

data=dsSimulate(s,'time_limits',[0 200],'solver','rk1','dt',.01,'mex_flag',1);
dsPlot(data);
```

**Comment by davestanley** [comment]
To fix this, I'm commenting out the lines

```
      fprintf(fid,'  if ~exist(''%s'',''var'')\n', condVariableName);
      fprintf(fid,'    %s;\n',initialization);
      fprintf(fid,'  end;\n');
```
from `dsWriteDynaSimSolver.m`

This will remove the following sections of the solve file:
```
if ~exist('I_V_spikes','var')
    I_V_spikes = [];
  end;
```

I have tested this and it appears to be okay (at least for the demo example) because I_V_spikes is defined earlier on in the file, under the monitors definition section:

```
% MONITORS:
E_tspike = -1e6*ones(2,p.E_Npop);
E_buffer_index = ones(1,p.E_Npop);
  E_V_spikes = zeros(nsamp,p.E_Npop);
I_tspike = -1e6*ones(2,p.I_Npop);
I_buffer_index = ones(1,p.I_Npop);
  I_V_spikes = zeros(nsamp,p.I_Npop);
```

**Comment by davestanley** [comment]
Just pushed this fix. @jsherfey I tested this change with the demo code, and it seemed to work. Hopefully it does not break other use cases (e.g., I_tspike should always be pre-defined in the monitors section)

**Comment by kupiqu** [comment]
This piece of code shouldn't be removed, as it it will break Octave, when I_V_spikes is not defined (issue #454), which used to be the case before monitoring the variable in dsDemos and can happen in general in other pieces of code.

Please revert. The proper way to go is doing as the error message indicates:

```
??? Function 'exist' is not supported for code generation. Consider adding coder.extrinsic('exist') at the top of the function to bypass code
generation.
```

**Comment by kupiqu** [comment]
Just want the let you know that I'm working on this...

**Comment by kupiqu** [comment]
Fixed with pull request #547

Unfortunately, not even coder.extrinsic('exist') works well with the Matlab Coder, so I had to workaround the issue in a way that I don't quite like because as mentioned in issue #454 Matlab's behavior here is very weird returning variables that do not even exist.

In any case, it works now in Matlab (with and without mex files), as well as in Octave.
